whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,398 @@
+Left "parser failed on []"
+```
+
+### Implementing a Alternative for our parser
+
+
+First we need a `Functor` instance for our `Parser i a` type which is pretty
+straight forward reusing the `Functor` instance for our inner `Maybe a`.
+
+```haskell
+{-# LANGUAGE InstanceSigs  #-}
+
+instance Functor (Parser i) where
+    fmap :: (a -> b) -> Parser i a -> Parser i b
+    fmap f (Parser p) =
+        Parser $ \is ->
+            let (aM, rest) = p is -- run the parser
+            in (fmap f aM, rest)  -- reuse the functor instance on the 'Maybe a`
+```
+
+Note: Above I've expicitly written the `fmap` function signature in the instance
+decleration which isn't required but makes it more easy to understand what
+is going on - for that you need to enable the `InstanceSigs` language extension.
+
+Next, before we can implement `Alternative` we also need to implement
+`Applicative` for our `Parser i a`:
+
+```haskell
+instance Applicative (Parser i) where
+    pure :: a -> Parser i a
+    pure a = Parser (\is -> (Just a, is))
+    -- with TupleSections could be written as : Parser (Just a,)
+
+    (<*>) :: Parser i (a -> b) -> Parser i a -> Parser i b
+    (Parser f) <*> (Parser p) = Parser $ \is ->
+      let (fM, rest) = f is
+      in case fM of
+           Nothing -> (Nothing, rest)
+           Just f' -> case p rest of
+                        (Just o, rest')  -> (pure $ f' o, rest')
+                        (Nothing, rest') -> (Nothing, rest')
+```
+
+I won't go into the details of `Applicative` here, as there are excelent other
+articles on the idea and ues of applicative functors. But with that beeing
+defined we are finally able to define the `Alternative` typeclass for our parser.
+
+
+```haskell
+instance Alternative (Parser i) where
+    empty = Parser (\is -> (Nothing, is))
+    -- with TupleSections could be written as: Parser (Nothing,)
+
+    (<|>) :: Parser i a -> Parser i a -> Parser i a
+    (Parser pa) <|> (Parser pb) = Parser $ \is ->
+      let (aMa, restA) = pa is
+          (aMb, restB) = pb is
+      in case (aMa, aMb) of
+           (Just a, _)       -> (Just a, restA)
+           (Nothing, Just b) -> (Just b, restB)
+           _                 -> (Nothing, is)
+```
+
+But why should we actually want an `Alternative` instance for a parser?
+First of all becuase of the possibilty to compose
+`Parser i a` with the `<|>` operator.
+Another reason are those two type class functions we haven't looked at so far
+but which we get for free:
+
+* `some :: f a -> f [a]` which means: _one or more_
+* `many :: f a -> f [a]` which means: _zero of more_
+
+So lets try:
+
+```none
+λ> evalParser (some $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (many $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (some $ parseLT 10) [10,2,3,20]
+Left "parser failed on [10,2,3,20]"
+
+λ> evalParser (many $ parseLT 10) [10,2,3,20]
+Right []
+```
+
+`some` and `many` are a very common and useful pattern in parser combinators.
```

Comment:
Can add a link to: https://serokell.io/blog/parser-combinators-in-haskell

Reply:
done